src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/ContainerRegistryTemplateProvider.cs

```diff
@@ -0,0 +1,105 @@
+        /// <returns>Template collection.</returns>
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(string templateCollectionReference, CancellationToken cancellationToken)
+        {
+            // We have embedded a default template set in the templatemanagement package.
+            // If the template set is the default reference, we don't need to retrieve token.
+            var registryServer = ExtractRegistryServer(templateCollectionReference);
+            var accessToken = string.Empty;
+            if (!IsDefaultTemplateReference(templateCollectionReference))
+            {
+                _logger.LogInformation("Using a custom template set for data conversion.");
+                accessToken = await _containerRegistryTokenProvider.GetTokenAsync(registryServer, cancellationToken);
+            }
+            else
+            {
+                _logger.LogInformation("Using the default template set for data conversion.");
+            }
+
+            try
+            {
+                var provider = _templateSetProviderFactory.CreateTemplateSetProvider(templateCollectionReference, accessToken);
+                return await provider.GetTemplateSetAsync(cancellationToken);
+            }
+            catch (ContainerRegistryAuthenticationException authEx)
+            {
+                _logger.LogError(authEx, "Failed to access container registry.");
+                throw new ContainerRegistryNotAuthorizedException(string.Format(Resources.ContainerRegistryNotAuthorized, registryServer), authEx);
+            }
+            catch (ImageFetchException fetchException)
+            {
+                _logger.LogError(fetchException, "Failed to fetch the templates from remote.");
+                throw new FetchTemplateCollectionFailedException(string.Format(Resources.FetchTemplateCollectionFailed, fetchException.Message), fetchException);
+            }
+            catch (ImageValidationException validationException)
+            {
+                _logger.LogError(validationException, "Failed to validate the downloaded image.");
+                throw new FetchTemplateCollectionFailedException(string.Format(Resources.FetchTemplateCollectionFailed, validationException.Message), validationException);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Unhandled exception: failed to get template set.");
+                throw new FetchTemplateCollectionFailedException(string.Format(Resources.FetchTemplateCollectionFailed, ex.Message), ex);
+            }
+        }
+
+        /// <summary>
+        /// Extract registry server from image reference (has been validated in the controller).
+        /// </summary>
+        /// <param name="imageReferece">image reference</param>
+        /// <returns>container registry server</returns>
+        private string ExtractRegistryServer(string imageReferece)
```

Comment:
>ExtractRegistryServer [](start = 23, length = 21)

I see another method with same name in DataConvertContoller.
Can `RegistryServer` from `DataConvertRequest` be used here instead of parsing that thing again?

Reply:
Fixed, thank you